### PR TITLE
[Feature] Allow `oneOf` in JSON schemas (with limited support)

### PR DESCRIPTION
**Pull request by hudson-ai**
- `oneOf` is allowed when only a single schema is provided in the `oneOf` list
   - should be uncontroversial
- Fall-back to `anyOf` if multiple schemas are provided, raising a warning to the user.
   - @riedgar-ms has already [voiced direct opposition to this](https://github.com/guidance-ai/guidance/issues/888#issuecomment-2152962427), but I think it warrants more discussion. In particular, the user may be using a JSON schema found "in the wild" and it may be difficult to reach down inside of them and "fix them" according to our limitations. Furthermore, I suspect that treating `oneOf` as `anyOf` will not cause any problems in a majority of use-cases. @Harsha-Nori do you have any opinions here?
- Add `id` and `discriminator` to ignored keys, expanding the schemas we support.
   - I *believe* this is without consequence, but I am not 100% sure for the discriminator case. Take a look at the [pydantic documentation for discriminated unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions). Note that the model in the example provided on this page uses both `oneOf` and `discriminator` when converted to a JSON schema.

### CODE REVIEW

**Comment by riedgar-ms** [comment]
As a side note, we can at least have our tests check that the correct warning was issued, but doing this does still make me a bit queasy.

File: tests/unit/library/test_json.py [hunk 1]
```diff
@@ -1253,6 +1253,37 @@ def test_allOf_bad_schema(self):
...
+        """
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=target_obj, schema=schema_obj)
+
+        # The actual check; we expect a warning here because oneOf is not fully supported
+        with pytest.warns() as record:
+            generate_and_check(target_obj, schema_obj)
+        assert len(record) == 1
+        assert record[0].message.args[0].startswith("oneOf not fully supported")
```
Comment (riedgar-ms) [inline tests/unit/library/test_json.py:1285]: Thanks for including this check (wrote my 'main' comment before seeing the code).

File: guidance/library/_json.py [hunk 2]
```diff
@@ -55,12 +57,14 @@ class Keyword(str, Enum):
 IGNORED_KEYS = {
     "$schema",
     "$id",
+    "id",
```
Comment (riedgar-ms) [inline guidance/library/_json.py:60]: I believe that our problem with `id` (in the FHIR schema) is only at top level. I think that `id` is a perfectly fine name for a property on an object.

File: guidance/library/_json.py [hunk 3]
```diff
@@ -55,12 +57,14 @@ class Keyword(str, Enum):
...
     "$comment",
     "title",
     "description",
     "default",
     "examples",
     "required",  # TODO: implement and remove from ignored list
+    "discriminator", # TODO: alternatively we could implement this in a stateful way
```
Comment (riedgar-ms) [inline guidance/library/_json.py:67]: MIght want to add a link to:
https://json-schema.org/blog/posts/validating-openapi-and-json-schema
which mentions how `discriminator` is part of the OpenAPI dialect of JSON-Schema.

**Review by riedgar-ms — CHANGES_REQUESTED** [review]
I don't think that `id` should be ignored completely. I think that
```json
{
    "type": "object",
    "properties": {
        "id" : {"type": "integer"}
    }
}
```
should be a perfectly valid schema.

File: guidance/library/_json.py (hunk 2 above)
Comment (hudson-ai) [reply guidance/library/_json.py:60]: `validate_json_node_keys` is only being called on dictionaries that represent full (sub)schemas, which doesn't include the dictionary specified by the `properties` key of a (sub)schema. Therefore the set of ignored keys should have no impact on what property names are valid -- worth checking if you have doubts :)

Comment (riedgar-ms) [reply guidance/library/_json.py:60]: I'd suggest adding a test with an object with an `id` property.

Comment (hudson-ai) [reply guidance/library/_json.py:60]: Happy to do so :)

File: tests/unit/library/test_json.py [hunk 4]
```diff
@@ -1981,3 +1981,13 @@ def test_no_additionalProperties(self, compact):
             maybe_whitespace=True,
             compact=compact,
         )
+
+def test_ignored_keys_allowed_as_properties():
```
Comment (hudson-ai) [inline tests/unit/library/test_json.py:1985]: @riedgar-ms is this test acceptable to you here? It doesn't *explicitly* test `id`; rather it checks against all ignored keys. Happy to specialize it to `id` if you like that better.

File: guidance/library/_json.py (hunk 3 above)
Comment (hudson-ai) [reply guidance/library/_json.py]: Done. Additionally added a longer comment about how disambiguating the grammar on this field could possibly improve performance or quality.

File: tests/unit/library/test_json.py (hunk 4 above)
Comment (riedgar-ms) [reply tests/unit/library/test_json.py:1985]: Even better to check them all

**Review by riedgar-ms — APPROVED** [review]

File: guidance/library/_json.py (hunk 3 above)
Comment (mmoskal) [reply guidance/library/_json.py]: If you have a discriminator, oneOf is equivalent to anyOf (since options are exclusive).

Comment (hudson-ai) [reply guidance/library/_json.py]: Nice observation and lets us more selectively warn the user. Might have to check that the discriminator is actually unique, but otoh maybe it's just up to the user to actually give us a valid schema...

**Pull request merged**